src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -110,12 +111,37 @@ public bool IsProcessingEnumerable()
             return (JsonTypeInfo.PropertyInfoForTypeInfo.ConverterStrategy & ConverterStrategy.Enumerable) != 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void MarkRequiredPropertyAsRead(JsonPropertyInfo propertyInfo)
         {
             if (propertyInfo.IsRequired)
             {
                 Debug.Assert(RequiredPropertiesLeft != null);
-                RequiredPropertiesLeft.Remove(propertyInfo);
+                RequiredPropertiesLeft.Remove(propertyInfo.Index);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void InitializeRequiredProperties(JsonTypeInfo typeInfo)
+        {
+            if (typeInfo.RequiredPropertiesIndices != null)
+            {
+                Debug.Assert(RequiredPropertiesLeft == null);
+                RequiredPropertiesLeft = new HashSet<int>(typeInfo.RequiredPropertiesIndices);
```

Comment:
After this refactoring, it's clear we could replace this with a `BitArray` or `bool[]` for a less compact representation.

Reply:
but then how do you efficiently check if all values are true/false? With HashSet I can simply check for Count

Reply:
In principle I agree backing up to 32/64 properties with an uint/UInt64 is a good and then falling back to something else but that would require testing, including having class with 64+ properties (or generated through contract customization)

Reply:
using BitArray now